### DISCUSSION TITLE
“Fatal glibc error "cannot get entropy for arc4random"” and Docker

### DISCUSSION THREAD

**Discussion by alexisLefebvre** [photoprism/photoprism#4550] [General]
I don't want to interfere with https://github.com/photoprism/photoprism/issues/4339 so I'm asking the question here.

Quote from @update-freak (and friendly ping)

> #### 4. What could be the cause of your problem?
> 
> I think the root cause is a too old kernel version. This has also been reported on other software: [jlesage/docker-freefilesync#8 (comment)](https://github.com/jlesage/docker-freefilesync/issues/8#issuecomment-1696377382), [MatthewVance/unbound-docker#147](https://github.com/MatthewVance/unbound-docker/issues/147))

Has anyone tried to report the issue to the Docker project directly? It affected several projects, so it may be an issue in Docker itself. Because in theory, the calls from any Docker container should be independent from the host system, yet it causes an issue here. So it looks like the layer between the container and the host has issue when `arc4random` is called.

What is the output of `docker --version` when it is run on a Synology device?

I'm actually surprised that people are able to run Docker and PhotoPrism on a Linux kernel from 2017 without having more issues.

**Comment by srett** [comment]
Docker is a container solution, it uses the kernel of the host system for syscalls, so if that one's too old, you'll encounter this issue.

**Comment by lastzero** [comment]
Based on my own research (in the glibc developer mailing list and source code), the error is caused by recent changes in glibc as they've changed the randomness generator without implementing a fallback for older kernels. Instead they throw a hard error or panic.

This essentially affects all applications that use glibc to generate random numbers / strings on an older kernel. It should be fixable, but from what I understand, they either forgot about it (as it was suggested during the initial discussion) or there simply wasn't time / funding for implementing a fallback.

I don't believe at this time that there are technical reasons for this, except maybe security considerations. However, arc4random() probability shouldn't be used for [cryptographic applications](https://security.stackexchange.com/questions/85601/is-arc4random-secure-enough) either, although the glibc manual states otherwise i.e. it could depend on the specific implementation and kernel version you use!?

**Reply by lastzero** [reply]
Related Discussions:
- https://lore.kernel.org/all/20220726195822.1223048-1-Jason@zx2c4.com/t/
- https://lore.kernel.org/all/?q=arc4random
